File: Partie1/minicipher.py
#!/usr/bin/python
# -*- coding: latin-1 -*-

########################################################
# minicipher.py : texte à trou pour la première partie #
########################################################


###############
# Chiffrement #
###############


# Définition des S-Box
s = [0xe,0x4,0xd,0x1,0x2,0xf,0xb,0x8,
     0x3,0xa,0x6,0xc,0x5,0x9,0x0,0x7]



# Definition de la permutation
perm = [0,4,8,12,
        1,5,9,13,
        2,6,10,14,
        3,7,11,15]



# Exécution d'une étape de chiffrement (à faire)
def encrypt_round (input, key, do_perm = True):
   """encrypt_round
   Paramètres
   ----------
   input   : bloc en entrée de l'étape
   key     : sous-clé de l'étape
   do_perm : si True, indique qu'il faut exécuter la permutation
             (permet de gérer la dernière étape sans permutation)
   
   Sortie
   ------
   la valeur de sortie de l'étape après calcul
   """
   
   # Addition de la sous-clé d'étape
   prelim = input^key 
   
   # Application des 4 S-Boxes
   inter = s[ (prelim >> 12) ] << 12 
   inter += s[ (prelim >> 8) & 0xF ] << 8
   inter += s[ (prelim >> 4) & 0xF ] << 4
   inter += s[ prelim & 0xF ] 
   
   # Si on applique pas la permutation, on peut rendre le résultat
   # tout de suite
   
   result = inter
   # Sinon, on applique la permutation, en envoyant chaque bit de
   # la sortie des S-Boxes vers sa destination réelle
   if(do_perm):
      result = 0
      for i in range(len(perm)):
         result += ((inter >> i) & 1) << perm[i]
   return result




# Chiffrement complet (à faire) #
def encrypt (plaintext, keys):
   """encrypt
   Paramètres
   ----------
   plaintext : texte clair en entrée
   keys      : tableau contenant les 5 sous-clés (i.e. la clé complète)
   
   Sortie
   ------
   la valeur du chiffré
   """
   NUM_ROUNDS = 4
   
   result = plaintext
   for i in range(NUM_ROUNDS):
      result = encrypt_round(result, keys[i], i != NUM_ROUNDS-1)
      
   result ^= keys[NUM_ROUNDS] 

   return result


#################
# Déchiffrement #
#################


# Déclaration des opérations inverses des S-Box
# et de la permutation
s_inv = [0] * 16
perm_inv = [0] * 16


# Définition des opérations inverses des S-Box
# et de la permutation (à faire)

def init_inverse_ops ():
   """Cette fonction remplit s_inv et perm_inv pour permettre le
      déchiffrement.
      Il faudra penser à appeler cette fonction AVANT toute
      opération de déchiffrement.
   """

   # Il faut remplir s_inv et perm_inv pour permettre le
   # déchiffrement
   for i in range(len(s)):
      s_inv[s[i]] = i

   for i in range(len(perm)):
      perm_inv[perm[i]] = i


# Exécution d'une étape de déchiffrement (à faire)
def decrypt_round (input, key, do_perm):
   """decrypt_round
   Paramètres
   ----------
   input   : bloc en entrée de l'étape
   key     : sous-clé de l'étape
   do_perm : si True, indique qu'il faut exécuter la permutation
             (permet de gérer la première étape sans permutation)
   
   Sortie
   ------
   la valeur de sortie de l'étape après calcul
   """
   post_perm = input
   if(do_perm):
      post_perm = 0
      for i in range(len(perm_inv)):
         post_perm += ((input >> i) & 1) << perm_inv[i]

   post_sub = s_inv[ (post_perm >> 12) ] << 12 
   post_sub += s_inv[ (post_perm >> 8) & 0xF ] << 8
   post_sub += s_inv[ (post_perm >> 4) & 0xF ] << 4
   post_sub += s_inv[ post_perm & 0xF ]

   result = post_sub ^ key

   return result

# Déchiffrement complet (à faire)
def decrypt (ciphertext, keys):
   """decrypt
   Paramètres
   ----------
   ciphertext : texte chiffré en entrée
   keys       : tableau contenant les 5 sous-clés (i.e. la clé complète)
   
   Sortie
   ------  
   la valeur du clair
   """

   NUM_ROUNDS = 4
   
   result = ciphertext ^ keys[NUM_ROUNDS] 

   for i in range(NUM_ROUNDS):
      result = decrypt_round(result, keys[-2-i], i != 0)

   return result

File: partie3/minicipher.c
/*******************************************************
 * minicipher.c : texte à trou pour la première partie *
 *******************************************************/



/***************
 * Chiffrement *
 ***************/


/* Définition des S-Box */
/************************/

int s[16] = {0xe,0x4,0xd,0x1,0x2,0xf,0xb,0x8,
	     0x3,0xa,0x6,0xc,0x5,0x9,0x0,0x7};



/* Définition de la permutation */
/********************************/

int perm[16] = {0,4,8,12,
		1,5,9,13,
		2,6,10,14,
		3,7,11,15};



/* Exécution d'une étape de chiffrement (à faire) */
/**************************************************/

int encrypt_round (int input, int key, int do_perm) {
  /* Paramètres
     ----------
  input   : bloc en entrée de l'étape
  key     : sous-clé de l'étape
  do_perm : si différent de zéro, indique qu'il faut exécuter la
            permutation (permet de gérer la dernière étape sans
            permutation

     Sortie
     ------
  la valeur de sortie de l'étape après calcul
  */

  int temp, result, i, p;

  // Addition de la sous-clé d'étape
  temp = input ^ key;

  // Application des 4 S-Boxes
  result = 0;
  for (i=0; i<16; i+=4)
    result |= (s[(temp >> i) & 0xf] << i);

  // Si on applique pas la permutation, on peut rendre le résultat
  // tout de suite
  if (do_perm == 0)
    return result;

  // Sinon, on applique la permutation, en envoyant chaque bit de
  // la sortie des S-Boxes vers sa destination réelle
  temp = result;
  result = 0;
  p = 1;
  for (i=0; i<16; i++, p<<=1)
    if (temp & p)
      result |= (1 << perm[i]);

  return result;
}



/* Chiffrement complet (à faire) */
/*********************************/

int encrypt (int plaintext, int keys[]) {
  /* Paramètres
     ----------
  plaintext : texte clair en entrée
  keys      : tableau contenant les 5 sous-clés (i.e. la clé complète)

     Sortie
     ------
  la valeur du chiffré
  */

  int i;
  int temp = plaintext;

  // Application des 4 étapes de chiffrement
  // (attention à ne pas effectuer la dernière permutation)
  for (i=0; i<4; i++)
    temp = encrypt_round (temp, keys[i], i!=3);
  
  // Enfin, on ajoute la dernière sous-clé à la fin
  temp ^= keys[4];

  return temp;
}






/*****************
 * Déchiffrement *
 *****************/


/* Déclaration des opérations inverses des S-Box */
/* et de la permutation                          */
/*************************************************/

int s_inv[16];
int perm_inv[16];


/* Définition des opérations inverses des S-Box */
/* et de la permutation (à faire)               */
/*************************************************/

void init_inverse_ops () {
  // Cette fonction remplit s_inv et perm_inv pour permettre le
  // déchiffrement
  // Il faudra penser à appeler cette fonction AVANT toute
  // opération de déchiffrement


  int i;
  for (i=0; i<16; i++) {
    s_inv[s[i]]=i;
    perm_inv[perm[i]]=i;
  }
}



/* Exécution d'une étape de déchiffrement (à faire) */
/****************************************************/

int decrypt_round (int input, int key, int do_perm) {
  /* Paramètres
     ----------
  input   : bloc en entrée de l'étape
  key     : sous-clé de l'étape
  do_perm : si différent de zéro, indique qu'il faut exécuter la
            permutation (permet de gérer la première étape sans
            permutation)

     Sortie
     ------
  la valeur de sortie de l'étape après calcul
  */


  int temp, result, i, p;

  // On commence par appliquer la permutation si cela est demandé
  if (do_perm != 0) {
    p = 1;
    temp = 0;
    for (i=0; i<16; i++, p<<=1)
      if (input & p)
	temp |= (1 << perm_inv[i]);
  } else {
    temp = input;
  }

  // Ensuite, on applique les S-Boxes inverses
  result = 0;
  for (i=0; i<16; i+=4)
    result |= (s_inv[(temp >> i) & 0xf] << i);
  
  // Addition de la sous-clé d'étape
  result ^= key;

  return result;
}



/* Déchiffrement complet (à faire) */
/***********************************/

int decrypt (int ciphertext, int keys[]) {
  /* Paramètres
     ----------
  ciphertext : texte chiffré en entrée
  keys       : tableau contenant les 5 sous-clés (i.e. la clé complète)

     Sortie
     ------  
  la valeur du clair
  */

  int i;
  int temp = ciphertext;

  // On commence par appliquer la dernière sous-clé
  temp ^= keys[4];

  // Application des 4 étapes de déchiffrement
  // (attention à ne pas effectuer la première permutation)
  for (i=0; i<4; i++)
    temp = decrypt_round (temp, keys[3-i], i!=0);

  return temp;
}





